crates/interpreter/src/exec.rs

```diff
@@ -1032,45 +1020,41 @@ impl<'m> Thread<'m> {
         self.values().split_off(len)
     }
 
-    fn push_label(&mut self, type_: FuncType<'m>, kind: LabelKind) {
-        let arity = match kind {
-            LabelKind::Block | LabelKind::If => type_.results.len(),
-            LabelKind::Loop => type_.params.len(),
-        };
-        let label = Label { arity, values_cnt: type_.params.len() };
-        self.label().values_cnt -= label.values_cnt;
-        self.labels().push(label);
+    fn push_label(&mut self) {
+        self.frame().labels_cnt += 1;
     }
 
     fn pop_label(mut self, l: LabelIdx, offset: usize) -> ThreadResult<'m> {
-        let i = self.labels().len() - l as usize - 1;
+        let frame = self.frame();
+        let i = frame.labels_cnt - l as usize - 1;
         if i == 0 {
             return self.exit_frame();
         }
-        let frame = self.frame();
-        let values_cnt: usize = frame.labels[i ..].iter().map(|label| label.values_cnt).sum();
-        let Label { arity, .. } = frame.labels.drain(i ..).next().unwrap();
+        frame.labels_cnt = i;
+        let side_table_entry = frame.side_table[offset].view();
         let values_len = self.values().len();
```

Comment:
This would simplify next line.
```suggestion
        let values_len = self.values().len() - side_table_entry.val_cnt as usize;
```